fix(cache): use `pathsCache` right after a directory has been visited

frantic1048 opened the pull request:
This fixes #52, see #52 for detail.

frantic1048 commented on index.js:
```diff
@@ -46,31 +46,34 @@ CaseSensitivePathsPlugin.prototype.reset = function() {
     return;
   }
+
+  this.fsOperations += 1;
```
I moved `fsOperations` counting after cache hit, to make it more sensible.

frantic1048 replied:
Resolved: outdated.

frantic1048 commented on index.js:
```diff
@@ -98,28 +101,30 @@ CaseSensitivePathsPlugin.prototype.fileExistsWithCase = function(
-          correctFilename = `\`${filenames[i]}\`.`;
-          break;
+  that.getFilenamesInDir(dir, (filenamesPromise) => {
+    filenamesPromise.then(filenames => {
```
Hidding whitespace changes is recommended, there is only indent change inside this code block.

frantic1048 replied:
Resolved: outdated.

frantic1048 commented on index.js:
```diff
@@ -46,31 +46,34 @@ CaseSensitivePathsPlugin.prototype.reset = function() {
-          dir,
-          err,
-        );
+  callback(new Promise(resolve => {
```
`getFilenamesInDir()`'s callback now accepts `Promise<string[]>`(which was `string[]`)

frantic1048 replied:
Resolved: outdated.

frantic1048 commented:
update: fixed test where multile callback is invoked on some case

Urthen commented:
Hi there,
Thanks for the contribution. I like the idea of speeding it up in general through parallelization but this isn't currently passing the tests. It looks like it's not properly failing if the folder case is wrong, at least that's the test failing on my machine.

Please make sure the tests pass locally and update.

frantic1048 commented on index.js:
```diff
@@ -29,54 +29,104 @@
+  //          causes as much as calls to fs.readdir()
+  // Thus we have to put a mutex for path `tippy`, when cache is not ready,
+  // to ensure we do not call fs.readdir() on same path multiple times.
+  this.visitedPathMutexRecord = new Map();
```
Here is why we need a mutex for each path.(described in comment)

Urthen replied:
Rather than having a mutex + cache be separate, I think this could be simplified by storing the promise in the `pathCache` as soon as the first request for a directory is made, then subsequent requests will see the promise and not make their own `fs.readdir` calls and instead return the same promise.

Does that make sense? That way we can also avoid adding a dependency; I've been trying to keep this package as intentionally light as possible.

frantic1048 commented on index.js:
```diff
@@ -29,54 +29,104 @@
   this.pathCache = new Map();
+  this.visitedPathMutexRecord.clear();
   this.fsOperations = 0;
-  this.primed = false;
```
`primed` is no longer used because we already make use of cache in `getFilenamesInDir()`.

frantic1048 commented on index.js:
```diff
@@ -29,54 +29,104 @@
+      that.logger.log(
+        '[CaseSensitivePathsPlugin] Hit cache for directory',
+        dir,
+      );
```
Logging for cache hit is added.

frantic1048 commented on index.js:
```diff
@@ -29,54 +29,104 @@
-  fs.readdir(dir, (err, files) => {
-    if (err) {
+  this.visitedPathMutexRecord.get(dir).runExclusive(() => {
+    if (that.pathCache.has(dir)) {
```
We may have several queued `runExclusive()` jobs when calling `getFilenamesInDir` with the same dir during a short while, after the first one writes the cache, the later ones should use the cached result.

Thus we have to check the cache first.

frantic1048 commented on index.js:
```diff
@@ -88,9 +138,10 @@ CaseSensitivePathsPlugin.prototype.fileExistsWithCase = function(
+    parsedPath.dir === parsedPath.root
+    || dir === '.'
+    // TODO: alternative way to express "known good path"
+    // || that.pathCache.has(filepath)
```
@Urthen I'm not very sure if removing this line will cause any (noticeable) impact on performance.

`pathCache` is now efficiently used to prevent redundant calls to `fs.readdir()`. Forgetting(not memorying) known good paths causes more checks.

On my machine, the overall result is positive on performance.

However, if this becomes an issue we could do further work to memorize known good path.

How do you think about this?

Once we get a resolution I will change this line to be more readable.

Urthen replied:
As far as I recall the `pathCache` needed to be primed with the current directory, and we needed to assume anything in the cache was correct. There was an issue where on windows you could CD into the directory with an incorrect case. While windows would allow it and otherwise there were no problems, when we were checking the case all the way down to root we would detect the incorrect case and fail the build unexpectedly. 

We'll need to keep the behavior of priming the cache & assuming the CWD is correctly cased in order to prevent a regression.

frantic1048 replied:
That's reasonable, I will try to keep the `primeCache()` logic and use Promise on `pathCache` to achieve it.

frantic1048 commented:
@Urthen Thanks for the inform. I revised the issue today and made a whole new fix. Of course, failing tests are fixed.

I updated the summary of changes in PR description, and added descriptions of key changes/questions in inline discussions.

Also, I resolved all outdated inline discussions opened by myself.

Review commits one by one is recommended since there is an ESlint styling fix commit.

Urthen reviewed (changes requested):
Thanks for all the work making this async so far, a few bits of feedback before this can get merged though. Once they're addressed I'll do some more testing on my own, then probably see if we can get some community testers to try it out as well.
